src/server/socket.js

```diff
@@ -1,115 +1,193 @@
+    subscribeTo(SIM_HOST, 'ready', handleSimHostReady, true);
+    emitTo(SIM_HOST, 'init');
+}
 
-            handlePendingEmits(APP_HOST);
-        });
+function onAppHostConnected() {
+    whenSimHostReady.promise.then(function () {
+        subscribeTo(APP_HOST, 'app-plugin-list', handleAppPluginList, true);
+        emitTo(APP_HOST, 'init');
+    });
+}
 
-        socket.on('register-simulation-host', function () {
-            log.log('Simulation host registered with server.');
+function onSimHostReady() {
+    setupAppHostHandlers();
+}
 
-            // It only makes sense to have one simulation host per server. If more than one tries to connect, always
-            // take the most recent.
-            hostSockets[SIM_HOST] = socket;
+function handleSimHostReady() {
+    // Resolve the deferred
+    whenSimHostReady.resolve();
 
-            socket.on('exec-success', function (data) {
-                emitToHost(APP_HOST, 'exec-success', data);
-            });
-            socket.on('exec-failure', function (data) {
-                emitToHost(APP_HOST, 'exec-failure', data);
-            });
+    setupSimHostHandlers();
 
-            socket.on('plugin-message', function (data) {
-                emitToHost(APP_HOST, 'plugin-message', data);
-            });
+    whenAppHostConnected.promise
+        .then(onAppHostConnected);
+}
 
-            socket.on('plugin-method', function (data, callback) {
-                emitToHost(APP_HOST, 'plugin-method', data, callback);
-            });
+function handleAppPluginList(data) {
+    whenSimHostReady.promise.then(function () {
+        subscribeTo(SIM_HOST, 'start', handleStart, true);
+        emitTo(SIM_HOST, 'app-plugin-list', data);
+    });
+}
 
-            socket.on('telemetry', function (data) {
-                telemetry.handleClientTelemetry(data);
-            });
+function handleStart() {
+    emitTo(APP_HOST, 'start');
+}
 
-            socket.on('debug-message', function (data) {
-                emitToHost(DEBUG_HOST, data.message, data.data);
-            });
+function setupSimHostHandlers() {
+    log.log('Setup handlers for SIM_HOST');
+
+    subscribeTo(SIM_HOST, 'exec-success', function (data) {
+        emitTo(APP_HOST, 'exec-success', data);
+    });
+
+    subscribeTo(SIM_HOST, 'exec-failure', function (data) {
+        emitTo(APP_HOST, 'exec-failure', data);
+    });
 
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
+    subscribeTo(SIM_HOST, 'plugin-message', function (data) {
+        emitTo(APP_HOST, 'plugin-message', data);
+    });
+
+    subscribeTo(SIM_HOST, 'plugin-method', function (data, callback) {
+        emitTo(APP_HOST, 'plugin-method', data, callback);
+    });
+
+    subscribeTo(SIM_HOST, 'debug-message', function (data) {
+        emitTo(DEBUG_HOST, data.message, data.data);
+    });
+
+    subscribeTo(SIM_HOST, 'telemetry', function (data) {
```

Comment:
I see that telemetry is sent over the socket even if not enabled, so please ignore my comment if it does not make sense.